### ISSUE TITLE
Wazuh Installation error in centos : Invalid user 'ossecm' or group 'ossec' given

### ISSUE THREAD

**Issue by HKcyberstark** [wazuh/wazuh#4816]
wazuh version : 3.12 | OS version : Centos 7
Component : Wazuh manager | Install type : both Packages and source script

Wazuh installation gets completed but service does not start.  `systemctl status wazuh-manager ` shows below error

```
[root@haran-ids-lab haran_kumar]# systemctl status wazuh-manager
● wazuh-manager.service - Wazuh manager
   Loaded: loaded (/etc/systemd/system/wazuh-manager.service; enabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Tue 2020-03-31 17:19:28 UTC; 7h ago
  Process: 5460 ExecStart=/usr/bin/env ${DIRECTORY}/bin/ossec-control start (code=exited, status=1/FAILURE)

Mar 31 17:19:28 haran-ids-lab systemd[1]: Starting Wazuh manager...
Mar 31 17:19:28 haran-ids-lab env[5460]: 2020/03/31 17:19:28 ossec-csyslogd: CRITICAL: (1203): Invalid user 'ossecm' or group 'ossec' given.
Mar 31 17:19:28 haran-ids-lab env[5460]: ossec-csyslogd: Configuration error. Exiting
Mar 31 17:19:28 haran-ids-lab systemd[1]: wazuh-manager.service: control process exited, code=exited status=1
Mar 31 17:19:28 haran-ids-lab systemd[1]: Failed to start Wazuh manager.
Mar 31 17:19:28 haran-ids-lab systemd[1]: Unit wazuh-manager.service entered failed state.
Mar 31 17:19:28 haran-ids-lab systemd[1]: wazuh-manager.service failed.
```

/etc/passwd  for ossec as below

```
ossec:x:994:991::/var/ossec:/sbin/nologin
ossecr:x:993:991::/var/ossec:/sbin/nologin
ossecm:x:992:991::/var/ossec:/sbin/nologin
```

**Comment by chrisboulton** [comment]
We're actually seeing something similar on some systems with 3.12 installed (about to try reproduce on 3.11) - instead of a username in the error though, it's just an empty string.

```
2020/04/01 22:29:55 ossec-execd: CRITICAL: (1203): Invalid user '' or group 'ossec' given.
```

I can see some calls in the code that trigger this error with an empty username, which seem to mostly be around group lookups. We've also confirmed the `ossec` group (and user) exists and is configured correctly.

(Debian Jessie here, Debian Buster seems unaffected from our testing)

**Comment by vikman90** [comment]
Hi @HKcyberstark @chrisboulton,

Thank you for opening this issue, I've taken a look at your thread in Slack. This error seems a bit sneaky, we've not managed to reproduce that yet.

I would like you to confirm that your file _/etc/group_ contains a line exactly like this:

```sh
# grep ossec /etc/group
ossec:x:991:
```

I've managed to reproduce it tampering that line manually.

On the other hand, you confirm that version 3.11 works in your system. The user/group management library was changed in 3.12 to fix a potential bug. Maybe that change is impacting your installation.

I've prepared this code to debug this problem, using both the old and the new implementation:

```c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <grp.h>
#include <assert.h>
#include <string.h>
#include <errno.h>

#define printerr(msg, ...) fprintf(stderr, msg, ##__VA_ARGS__)

static gid_t Privsep_GetGroup_new(const char * name)
{
    struct group grp = { .gr_name = NULL };
    struct group * result = NULL;
    gid_t gr_gid = (gid_t) -1;

    // Get the correct buffer size

    long int len = sysconf(_SC_GETGR_R_SIZE_MAX);
    printerr("DEBUG: sysconf(_SC_GETGR_R_SIZE_MAX) = %ld\n", len);
    len = len > 0 ? len : 1024;

    // Allocate buffer in the heap

    char * buffer = malloc(len);
    assert(buffer != NULL);

    // Retrieve the group data

    int retval = getgrnam_r(name, &grp, buffer, len, &result);
    printerr("DEBUG: getgrnam_r() = %d\n", retval);
    printerr("DEBUG: result = (%s object)\n", result == NULL ? "null " : "valid");

    // Check the result

    if (retval == 0) {
        if (result != NULL) {
            printerr("DEBUG: getgrnam_r(): group found.\n");
            gr_gid = result->gr_gid;
        } else {
            printerr("DEBUG: getgrnam_r(): group not found.\n");
        }
    } else {
        printerr("ERROR: getgrnam_r(): %s.\n", strerror(retval));
    }

    // Clean and return

    free(buffer);
    return gr_gid;
}

static gid_t Privsep_GetGroup_old(const char * name) {
    errno = 0;

    // Retrieve the group data
    struct group * result = getgrnam(name);

    // Check the result

    if (result != NULL) {
        printerr("DEBUG: getgrnam(): group found.\n");
        return result->gr_gid;
    } else {
        switch (errno) {
        case 0:
        case ENOENT:
        case ESRCH:
        case EBADF:
        case EPERM:
            printerr("DEBUG: getgrnam(): group not found: %s (%d)\n", strerror(errno), errno);
            break;

        default:
            printerr("DEBUG: getgrnam(): %s (%d).\n", strerror(errno), errno);
        }

        return -1;
    }
}

int main(int argc, char ** argv) {
    int return_code = 0;

    if (argc < 2) {
        printerr("ERROR: group name not given.\nSyntax: %s <group>\n", argv[0]);
        return EXIT_FAILURE;
    }

    // New implementation

    gid_t gid = Privsep_GetGroup_new(argv[1]);

    if (gid != (gid_t) -1) {
        printf("%d\n", (int)gid);
    } else {
        return_code = EXIT_FAILURE;
    }

    // Old implementation

    gid = Privsep_GetGroup_old(argv[1]);

    if (gid != (gid_t) -1) {
        printf("%d\n", (int)gid);
    } else {
        return_code = EXIT_FAILURE;
    }

    return return_code;
}
```

Could you compile and run this code in your testing environment?

This is the expected result (assuming the source code file is _getgrnam.c_):

```sh
$ gcc -o getgrnam getgrnam.c
$ ./getgrnam ossec
DEBUG: sysconf(_SC_GETGR_R_SIZE_MAX) = 1024
DEBUG: getgrnam_r() = 0
DEBUG: result = (valid object)
DEBUG: getgrnam_r(): group found.
991
DEBUG: getgrnam(): group found.
991
```

Please confirm that the  _ossec_ user is installed properly, and write back to us with the result of the testing program. That will help us to figure out the cause of this issue.

Best regards.

**Comment by chrisboulton** [comment]
Hey @vikman90 - thank you for digging into this.

```
$ ./test ossec
DEBUG: sysconf(_SC_GETGR_R_SIZE_MAX) = 1024
DEBUG: getgrnam_r() = 34
DEBUG: result = (null  object)
ERROR: getgrnam_r(): Numerical result out of range.
DEBUG: getgrnam(): group found.
124
```

/etc/group entry:

```
$ grep ossec /etc/group
ossec:x:124:
```

I just had a look at the usage of `getgrnam_r`, and I can see your test code is correctly making the `sysconf(_SC_GETGR_R_SIZE_MAX)` call to get the suggested buffer size, but it's just a suggestion and it looks like in our case we're exceeding it. Per the docs: 

> (If this size is too small, the call fails with ERANGE, in which case the caller can retry with a larger buffer.)

Sounds like that this particular case needs to be handled for with a retry - possibly just by incrementing by the `sysconf(..)` suggestion each time `ERANGE` is returned? (up to some safe limit, I guess)

**Comment by vikman90** [comment]
@chrisboulton you nailed it!

Thank you very much for running this program. Finally, I managed to reproduce the issue: we must manage groups with a very long name. This is a bit weird for me since `addgroup` does not allow more than 32 bytes long names.

## How to reproduce

1. Create a group with a short name. 
2. Replace that name with a very long word in _/etc/group_.
3. Run the program above.

```sh
$ groupadd abcdefghijklmnopqrstuvwxyzabcdef
$ sed -i "s/abcdefghijklmnopqrstuvwxyzabcdef/abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz/" /etc/group

$ ./getgrnam abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuv[root@centos7 ~]# sed -i "s/abcdefghijklmnopqrstufghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz/" /etc/group
DEBUG: sysconf(_SC_GETGR_R_SIZE_MAX) = 1024
DEBUG: getgrnam_r() = 34
DEBUG: result = (null  object)
ERROR: getgrnam_r(): Numerical result out of range.
DEBUG: getgrnam(): group found.
1001

$ groupdel abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz
```

## Proposed fix

We could let `sysconf()` define the initial size of the buffer, and make the program increase it up to 64 KiB if it's not enough, just as you suggested.

### Third implementation

```c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <grp.h>
#include <assert.h>
#include <string.h>
#include <errno.h>

#define printerr(msg, ...) fprintf(stderr, msg, ##__VA_ARGS__)

static gid_t Privsep_GetGroup_new_2(const char * name) {
    const long MAX_LEN = 65536;
    struct group grp = { .gr_name = NULL };
    struct group * result = NULL;
    gid_t gr_gid = (gid_t) -1;
    char * buffer = NULL;
    int retval;

    // Get the intiial buffer size

    long int len = sysconf(_SC_GETGR_R_SIZE_MAX);
    printerr("DEBUG: sysconf(_SC_GETGR_R_SIZE_MAX) = %ld\n", len);

    do {
        // Allocate buffer in the heap

        printerr("DEBUG: realloc(%ld)\n", len);
        buffer = realloc(buffer, len);
        assert(buffer != NULL);

        // Retrieve the group data

        retval = getgrnam_r(name, &grp, buffer, len, &result);
        printerr("DEBUG: getgrnam_r() = %d\n", retval);
        printerr("DEBUG: result = (%s object)\n", result == NULL ? "null " : "valid");

        // Retry if the buffer was not long enough
    } while (errno == ERANGE && (len *= 2) <= MAX_LEN);

    // Check the result

    if (retval == 0) {
        if (result != NULL) {
            printerr("DEBUG: getgrnam_r(): group found.\n");
            gr_gid = result->gr_gid;
        } else {
            printerr("DEBUG: getgrnam_r(): group not found.\n");
        }
    } else {
        printerr("ERROR: getgrnam_r(): %s.\n", strerror(retval));
    }

    // Clean and return

    free(buffer);
    return gr_gid;
}

static gid_t Privsep_GetGroup_new(const char * name)
{
    struct group grp = { .gr_name = NULL };
    struct group * result = NULL;
    gid_t gr_gid = (gid_t) -1;

    // Get the correct buffer size

    long int len = sysconf(_SC_GETGR_R_SIZE_MAX);
    printerr("DEBUG: sysconf(_SC_GETGR_R_SIZE_MAX) = %ld\n", len);
    len = len > 0 ? len : 1024;

    // Allocate buffer in the heap

    char * buffer = malloc(len);
    assert(buffer != NULL);

    // Retrieve the group data

    int retval = getgrnam_r(name, &grp, buffer, len, &result);
    printerr("DEBUG: getgrnam_r() = %d\n", retval);
    printerr("DEBUG: result = (%s object)\n", result == NULL ? "null " : "valid");

    // Check the result

    if (retval == 0) {
        if (result != NULL) {
            printerr("DEBUG: getgrnam_r(): group found.\n");
            gr_gid = result->gr_gid;
        } else {
            printerr("DEBUG: getgrnam_r(): group not found.\n");
        }
    } else {
        printerr("ERROR: getgrnam_r(): %s.\n", strerror(retval));
    }

    // Clean and return

    free(buffer);
    return gr_gid;
}

static gid_t Privsep_GetGroup_old(const char * name) {
    errno = 0;

    // Retrieve the group data
    struct group * result = getgrnam(name);

    // Check the result

    if (result != NULL) {
        printerr("DEBUG: getgrnam(): group found.\n");
        return result->gr_gid;
    } else {
        switch (errno) {
        case 0:
        case ENOENT:
        case ESRCH:
        case EBADF:
        case EPERM:
            printerr("DEBUG: getgrnam(): group not found: %s (%d)\n", strerror(errno), errno);
            break;

        default:
            printerr("DEBUG: getgrnam(): %s (%d).\n", strerror(errno), errno);
        }

        return -1;
    }
}

int main(int argc, char ** argv) {
    int return_code = 0;

    if (argc < 2) {
        printerr("ERROR: group name not given.\nSyntax: %s <group>\n", argv[0]);
        return EXIT_FAILURE;
    }

    // Old implementation

    printerr("==== First version =====\n");
    gid_t gid = Privsep_GetGroup_old(argv[1]);

    if (gid != (gid_t) -1) {
        printf("%d\n", (int)gid);
    } else {
        return_code = EXIT_FAILURE;
    }

    // New implementation

    printerr("\n==== Second version ====\n");
    gid = Privsep_GetGroup_new(argv[1]);

    if (gid != (gid_t) -1) {
        printf("%d\n", (int)gid);
    } else {
        return_code = EXIT_FAILURE;
    }

    // Third version

    printerr("\n==== Third version =====\n");
    gid = Privsep_GetGroup_new_2(argv[1]);

    if (gid != (gid_t) -1) {
        printf("%d\n", (int)gid);
    } else {
        return_code = EXIT_FAILURE;
    }

    return return_code;
}
```

### Expected output

```
==== First version =====
DEBUG: getgrnam(): group found.
1001

==== Second version ====
DEBUG: sysconf(_SC_GETGR_R_SIZE_MAX) = 1024
DEBUG: getgrnam_r() = 34
DEBUG: result = (null  object)
ERROR: getgrnam_r(): Numerical result out of range.

==== Third version =====
DEBUG: sysconf(_SC_GETGR_R_SIZE_MAX) = 1024
DEBUG: realloc(1024)
DEBUG: getgrnam_r() = 34
DEBUG: result = (null  object)
DEBUG: realloc(2048)
DEBUG: getgrnam_r() = 0
DEBUG: result = (valid object)
DEBUG: getgrnam_r(): group found.
1001
```

Would you be so kind as to launch the new version of the program? This way we will be sure it fixes the issue.

Out of curiosity, what is the longest group name in your system?

I appreciate your feedback.
Best regards.

**Comment by HKcyberstark** [comment]
Hello folks 
@vikman90   - what would be the perm fix for this  ?  I am having suricata, zeek in box and waiting for wazuh installation on same box. 
in the interim , how can I install with 3.10 which might be a stable version for my POC ?

**Comment by vikman90** [comment]
@HKcyberstark the fix requires us to apply the patch I proposed above. On the other hand, this problem seems to be due to having a group with a very long name in your host. Obviously, you should be able to create groups names as long as you want —and your host allows—, but if you could rename them provisionally, it could work.

You can use the [Wazuh 3.10 installation guide](https://documentation.wazuh.com/3.10/installation-guide/index.html) to install a previous version of Wazuh. Please note that, as you are experiencing this problem in a manager, you need to downgrade all your agents to 3.10.

We will continue testing the patch for this issue. If you guys could run the latest code I proposed, that would prove that the fix works for you.

Best regards.

**Comment by chrisboulton** [comment]
@vikman90 Sorry for the delay. The fix looks good - it works, and the code changes make sense to me.

```
==== First version =====
DEBUG: getgrnam(): group found.
125

==== Second version ====
DEBUG: sysconf(_SC_GETGR_R_SIZE_MAX) = 1024
DEBUG: getgrnam_r() = 34
DEBUG: result = (null  object)
ERROR: getgrnam_r(): Numerical result out of range.

==== Third version =====
DEBUG: sysconf(_SC_GETGR_R_SIZE_MAX) = 1024
DEBUG: realloc(1024)
DEBUG: getgrnam_r() = 34
DEBUG: result = (null  object)
DEBUG: realloc(2048)
DEBUG: getgrnam_r() = 0
DEBUG: result = (valid object)
DEBUG: getgrnam_r(): group found.
125
```

For some context, the longest group name length in our systems is only 18 characters. There's 80 groups, and a couple of groups have quite a few members.

**Comment by vikman90** [comment]
@chrisboulton thank you for testing the patch, that is a huge help to us!

I've ported this fix to the code (PR #4873). Unfortunately, 3.12.1 is closed now, but the fix should be available in the next version, hopefully, 3.12.2.

Hope this fix helps you.
Best regards.

**Comment by vikman90** [comment]
@chrisboulton @HKcyberstark We've just merged the fix into the branch 3.12, so it will apply in the next patch version, that we plan to release in about 2 weeks.

Thank you very much for your help with this issue. Let me close it as the fix is applied.

Best regards.
